ddl/check.lua

```diff
@@ -628,75 +628,190 @@ local function check_index(i, index, space)
+            space.name
+        )
+    end
+
+    if space.indexes.bucket_id.unique then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].unique: bucket_id index can't be unique",
+            space.name, 'bucket_id'
+        )
+    end
+
+    if #space.indexes['bucket_id'].parts ~= 1 then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts: bucket_id index can't be composite (1 part expected, got %d parts)",
+            space.name, 'bucket_id', #space.indexes['bucket_id'].parts
+        )
+    end
+
+    if space.indexes['bucket_id'].parts[1].path ~= 'bucket_id' then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts[1].path: invalid field reference " ..
+            "(reference to bucket_id expected, got %s)",
+            space.name, 'bucket_id', space.indexes['bucket_id'].parts[1].path
+        )
+    end
+
+    return true
+end
+
+local function is_callable(object)
+    if type(object) == 'function' then
+        return true
+    end
+
+    -- all objects with type `cdata` are allowed
+    -- because there is no easy way to get
+    -- metatable.__call of object with type `cdata`
+    if type(object) == 'cdata' then
+        return true
+    end
+
+    local object_metatable = getmetatable(object)
+    if (type(object) == 'table' or type(object) == 'userdata') then
+        -- if metatable type is not `table` -> metatable is protected ->
+        -- cannot detect metamethod `__call` exists
+        if object_metatable and type(object_metatable) ~= 'table' then
+            return true
         end
 
-        if space.indexes.bucket_id ~= nil then
-            return nil, string.format(
-                "spaces[%q].indexes[%q]: bucket_id is used for sharding, " ..
-                "but there's no spaces[%q].sharding_key defined",
-                space.name, 'bucket_id', space.name
-            )
+        -- `__call` metamethod can be only the `function`
+        -- and cannot be a `table` | `userdata` | `cdata`
+        -- with `__call` methamethod on its own
+        if object_metatable and object_metatable.__call then
+            return type(object_metatable.__call) == 'function'
         end
-        return true
     end
 
-    do -- check that format['bucket_id'] valid
-        if not space.fields.bucket_id then
-            return nil, string.format(
-                "spaces[%q].format: sharding_key exists in the space, but there's" ..
-                " no bucket_id defined in 'format' section",
-                space.name
-            )
+    return false
+end
+
+local function check_name_isident(name)
+    if name == nil or name == '' then
+        return false
+    end
+
+    -- sharding function name cannot
+    -- be equal to lua keyword
+    if utils.LUA_KEYWORDS[name] then
+        return false
+    end
+
+    -- sharding function name cannot
+    -- begin with a digit
+    if tonumber(name:sub(1,1)) ~= nil then
+        return false
+    end
```

Comment:
Nit: Since we already have the `lj_char_bits` table, we can define `lj_char_isdigit()`. Feel free to ignore, it is purely thing of taste (but looks logical for me, because we use `lj_char_isident()` just below).